Invalid C++ code generation when returning var T

For the following example, the code that is generated is wrong and induce a segmentation fault.

### Example 
```nim
type 
    Vector*  {.importcpp: "std::vector", header: "vector".}[T] = object

proc initVector*[T](n: csize): Vector[T] 
    {.importcpp: "std::vector<'*0>(@)", header: "vector".}

proc unsafeIndex[T](this: var Vector[T], i: csize): var T 
    {.importcpp: "#[#]", header: "vector".}

proc `[]`*[T](this: var Vector[T], i: Natural): var T {.inline, noinit.} =
    when compileOption("boundChecks"):
        # this.checkIndex i
        discard
    result = this.unsafeIndex(i)

var v1 = initVector[int](10)
echo v1[0] # <-- SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

The error comes from the following generated lines:
```
static N_INLINE(NI*, X5BX5D__9bJ9bA39bRmdl9bxuHGDHoW1tQplay)(TY_DImjm4JODOV5hEJosD0PGw& this_0, NI i) {
	NI* result;
	nimfr_("[]", "play.nim");
	nimln_(16, "play.nim");
	result = this_0[((size_t) (i))]; // <-- should be:  result = &this_0[((size_t)(i))];
	popFrame();
	return result;
}
```
### Current Output

```
Traceback (most recent call last)
play.nim(17)             play
SIGSEGV: Illegal storage access. (Attempt to read from nil?)

```

### Expected Output

```
0
```

### Workarounds

1. using this strange syntax fixes the bug:
```nim
proc `[]`*[T](this: var Vector[T], i: Natural): var T {.inline, noinit.} =
    when compileOption("boundChecks"):
        # this.checkIndex i
        discard
    # This strange syntax is to avoid a bug in the Nim c++ code generator
    result = (addr this.unsafeIndex(i))[]
```
Here is the code generated:
```
static N_INLINE(NI*, X5BX5D__9bJ9bA39bRmdl9bxuHGDHoW1tQplay)(TY_DImjm4JODOV5hEJosD0PGw& this_0, NI i) {
	NI* result;
	nimfr_("[]", "play.nim");
	nimln_(25, "play.nim");
	result = (&this_0[((size_t) (i))]); // <-- this is good
	popFrame();
	return result;
}
```

2. use a template instead of a proc


### Additional Information
nim --version
```
Nim Compiler Version 0.19.9 [Linux: amd64]
Compiled at 2019-01-03
Copyright (c) 2006-2018 by Andreas Rumpf

active boot switches: -d:release
```

Thank for your help. I tried using a pointer like you suggested because it seems to be the thing to do but the problem is that I get another invalid code generation error:

```nim
import strformat

type 
    Vector*  {.importcpp: "std::vector", header: "vector".}[T] = object

proc initVector*[T](n: csize): Vector[T] 
    {.importcpp: "std::vector<'*0>(@)", header: "vector".}

proc size*[T](this: Vector[T]): csize 
    {.noSideEffect, importcpp: "size", header: "vector".}

proc checkIndex[T](this: Vector[T], i: csize) : csize {.inline.} =
    if 0 > i or i > this.size:
        raise newException(IndexError, 
            &"index out of bounds: 0 <= (i:{i}) <= (n:{this.size})")
    result = i

proc unsafeIndex[T](this: var Vector[T], i: csize): var T 
    {.importcpp: "&#[#]", header: "vector".}

proc `[]`*[T](this: var Vector[T], i: Natural): var T {.inline, noinit.} =
    result = this.unsafeIndex(this.checkIndex i)

proc `[]=`*[T](this: var Vector[T], i: Natural, val: T) {.inline, noinit.} =
    this.unsafeIndex(this.checkIndex i) = val

var v1 = initVector[int](10)
v1[0] = 100 # <-- wrong code generated for this line
echo v1[0] 
```
which gives the following error:
```
/home/user/.cache/nim/play_d/play_play.cpp: In function ‘void X5BX5Deq__ZVQLhFsIfFuHeNr2UXhW2wplay(TY_DImjm4JODOV5hEJosD0PGw&, NI, NI)’:
/home/user/.cache/nim/play_d/play_play.cpp:443:74: error: lvalue required as left operand of assignment
  &this_0[checkIndex_USyPXguD9caC5SagGtT8zBAplay(this_0, ((size_t) (i)))] = val;
```

--- a/tests/cpp/torc.nim
+++ b/tests/cpp/torc.nim
@@ -39,3 +39,23 @@ proc asVector*[T](t: T): EnumVector[T] =
 # N_NIMCALL(std::vector<> , asvector_106028_3197418230)(SomeEnum t0)
 
 discard asVector(SomeEnum.A)
+
+
+block: # bug #10219
+  type
+    Vector[T]  {.importcpp: "std::vector", header: "vector".} = object
+
+  proc initVector[T](n: csize_t): Vector[T] 
+      {.importcpp: "std::vector<'*0>(@)", header: "vector".}
+
+  proc unsafeIndex[T](this: var Vector[T], i: csize_t): var T 
+      {.importcpp: "#[#]", header: "vector".}
+
+  proc `[]`[T](this: var Vector[T], i: Natural): var T {.inline, noinit.} =
+    when compileOption("boundChecks"):
+        # this.checkIndex i
+        discard
+    result = this.unsafeIndex(csize_t(i))
+
+  var v1 = initVector[int](10)
+  doAssert v1[0] == 0

